### ISSUE TITLE
How to make the motion of AI vehicle in Beamng obey the script node

### ISSUE THREAD

**Issue by sainan-zhang** [BeamNG/BeamNGpy#193]
Hi, In my scenario, there is only an AI vehicle and a controlled vehicle. I want to slow down the AI vehicle after it quickly overtook the controlled vehicle in Beamng. I tried to use the script node to control the movement of AI vehicle. But I found that even though I use the function 'time.sleep(0.1)', the speed of AI is low and the AI vehicle make strange noises before the overtaking . Besides, after the AI vehicle overtakes, it moves strangely, with a sinusoidal-like trajectory. But the script node I set is a straight-line trajectory after it overtook. I'm exceedingly confused. My code is shown below. Can you help me? Looking forward to your reply! Thanks very much!

def getAiScript0(orig=(0, -10, 0),tb=0):
    script = list()
    for i in range(3600):
        node = {
            'x': orig[0],
            'y': -i * 20 + orig[1],
            'z': orig[2],
            't': i+tb,
        }
        script.append(node) # drive with high speed, before overtaking
    return script 

def getAiScript2(orig=(0, -10, 0),tb=0):
    script = list()
    for i in range(3600):
        node = {
            'x': orig[0],
            'y': -i * 3 + orig[1],
            'z': orig[2],
            't': i+tb,
        }
        script.append(node) # drive with low speed,after overtaking
    return script 

def resetAiScript0(script, pos_x):
    for node in script:
        node['x'] = pos_x
    return script

beamng.open()
scenario = Scenario('smallgrid', 'test')
veh_cont = Vehicle('ego_vehicle', model='etk800', licence='control')
veh_auto1 = Vehicle('ai_vehicle1', model='etki', licence='auto', colour="Red")

electrics = Electrics()
electrics1 = Electrics()
veh_cont.attach_sensor('electrics', electrics)
veh_auto1.attach_sensor('electrics', electrics1)

timer = Timer()
timer1 = Timer()
veh_cont.attach_sensor('timer', timer)
veh_auto1.attach_sensor('timer', timer1)

damage = Damage()
damage1 = Damage()
veh_cont.attach_sensor('damage', damage)
veh_auto1.attach_sensor('damage', damage1)

posCon = (-2, -100, 0.40)
posAuto1 = (2, -100, 0.40)

scenario.add_vehicle(veh_cont, pos=(0, 0, 0.4), rot=(0, 0, 0))
scenario.add_vehicle(veh_auto1, pos=(0, -10, 0.4), rot=(0, 0, 0))

scenario.make(beamng)
beamng.load_scenario(scenario)

beamng.teleport_vehicle(veh_cont.vid, posCon, rot=(0, 0, 0))
beamng.teleport_vehicle(veh_auto1.vid, posAuto1, rot=(0, 0, 0))

beamng.switch_vehicle(veh_cont)
script1 = getAiScript0(posAuto1)

beamng.start_scenario()

veh_auto1.ai_set_script(script1)

while True:
    same1 = 0

    sensors = beamng.poll_sensors(veh_cont)
    veh_cont.update_vehicle()

    sensors1 = beamng.poll_sensors(veh_auto1)
    veh_auto1.update_vehicle()

    pos0 = veh_cont.state['pos']
    pos1 = veh_auto1.state['pos']

    timeNow = sensors['timer']['time']

    sameLine = (pos0[0] * pos1[0] > 0)

    if pos0[1] - pos1[1] > 10:
        if sameline == False:
            same1 = 0
            x_con = pos0[0]
            script11= getAiScript0(pos1, timeNow)
            script111 = resetAiScript0(script11, x_con)
            veh_auto1.ai_set_script(script111)
        else:
            script2 = getAiScript2(pos1, timeNow)
            if same1==0:
                veh_auto1.ai_set_script(script2)
                same1 = 1

    if pos0[1] - pos1[1] <= 10:
        if pos0[0] > 0:
            script3 = getAiScript0(pos1, timeNow)
            script33 = resetAiScript0(script3, -3)
            veh_auto1.ai_set_script(script33)
        else:
            script4 = getAiScript0(pos1,timeNow)
            script44 = resetAiScript0(script4, 3)
            veh_auto1.ai_set_script(script44)

    time.sleep(0.1)

    if timeNow > 300:
        break

**Comment by aivora-beamng** [comment]
Hi, unfortunately it's quite hard to follow your script and find out what you are trying to do exactly. Also it seems you encounter a multiple issues, it would be much easier to investigate if you try to create a smaller separate snippet of code for every issue you encounter.

I'll try to address the problems you seem to be having:
1. speed of AI is low before overtaking
    Not sure if you mean the response time of AI, or the speed of the AI vehicle. You can try increasing the time between sensor polling (`time.sleep(0.5)` for example) if you still experience the simulator lagging. Another possible reason could be your hardware is not sufficient enough.
2. after the AI vehicle overtakes, it moves strangely, with a sinusoidal-like trajectory
    This could be a bug on the simulation side. Could you try to isolate a smaller code example of when this behaviour happens and post it either here or as a new issue?
3. I want to slow down the AI vehicle
    You can use the function [Vehicle.ai_set_speed](https://beamngpy.readthedocs.io/en/stable/beamngpy.html#beamngpy.vehicle.Vehicle.ai_set_speed) for setting the target speed of the AI vehicle, like in this snippet:
```python
veh_auto1.ai_set_speed(10, mode='limit')
```
The speed is in m/s, and the mode can be either `limit` or `set`.
